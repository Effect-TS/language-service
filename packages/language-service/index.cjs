// eslint-disable-next-line @typescript-eslint/consistent-type-imports, no-undef
module.exports = function init(modules) {
  function create(info) {
    // start with regular language service, then load custom once async module has loaded
    let languageService = info.languageService
    import("./build/cjs/index.js").then(v => languageService = v.default(modules).create(info))

    // create the proxy
    const proxy = Object.create(null)
    for (const k of Object.keys(info.languageService)) {
      proxy[k] = (...args) => languageService[k].apply(languageService, args)
    }

    return proxy
  }

  return { create }
}
