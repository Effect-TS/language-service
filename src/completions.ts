import { contextSelfInClasses } from "./completions/contextSelfInClasses.js"
import { effectDataClasses } from "./completions/effectDataClasses.js"
import { effectSchemaSelfInClasses } from "./completions/effectSchemaSelfInClasses.js"
import { effectSelfInClasses } from "./completions/effectSelfInClasses.js"
import { pipeNonPipeables } from "./completions/pipeNonPipeables.js"

export const completions = [
  pipeNonPipeables,
  effectSchemaSelfInClasses,
  effectSelfInClasses,
  contextSelfInClasses,
  effectDataClasses
]
