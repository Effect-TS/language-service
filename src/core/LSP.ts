import * as ReadonlyArray from "effect/Array"
import * as Data from "effect/Data"
import { pipe } from "effect/Function"
import * as Option from "effect/Option"
import type ts from "typescript"
import type * as TypeCheckerApi from "../core/TypeCheckerApi.js"
import * as TypeScriptApi from "../core/TypeScriptApi.js"
import * as Nano from "./Nano.js"

export class RefactorNotApplicableError
  extends Data.TaggedError("RefactorNotApplicableError")<{}>
{}

export interface RefactorDefinition {
  name: string
  description: string
  apply: (
    sourceFile: ts.SourceFile,
    textRange: ts.TextRange
  ) => Nano.Nano<
    ApplicableRefactorDefinition,
    RefactorNotApplicableError,
    | TypeScriptApi.TypeScriptApi
    | TypeCheckerApi.TypeCheckerApi
    | PluginOptions
    | TypeCheckerApi.TypeCheckerApiCache
  >
}

export interface ApplicableRefactorDefinition {
  kind: string
  description: string
  apply: Nano.Nano<void, never, ts.textChanges.ChangeTracker>
}

export function createRefactor(definition: RefactorDefinition): RefactorDefinition {
  return definition
}

export interface DiagnosticDefinition {
  name: string
  code: number
  apply: (
    sourceFile: ts.SourceFile
  ) => Nano.Nano<
    Array<ApplicableDiagnosticDefinition>,
    never,
    | TypeCheckerApi.TypeCheckerApi
    | PluginOptions
    | TypeScriptApi.TypeScriptApi
    | TypeCheckerApi.TypeCheckerApiCache
  >
}

export interface ApplicableDiagnosticDefinition {
  node: ts.Node
  category: ts.DiagnosticCategory
  messageText: string
  fixes: Array<ApplicableDiagnosticDefinitionFix>
}

export interface ApplicableDiagnosticDefinitionFix {
  fixName: string
  description: string
  apply: Nano.Nano<void, never, ts.textChanges.ChangeTracker>
}

export interface ApplicableDiagnosticDefinitionFixWithPositionAndCode
  extends ApplicableDiagnosticDefinitionFix
{
  code: number
  start: number
  end: number
}

export function createDiagnostic(definition: DiagnosticDefinition): DiagnosticDefinition {
  return definition
}

export interface PluginOptions {
  diagnostics: boolean
  quickinfo: boolean
  completions: boolean
}

export interface CompletionDefinition {
  name: string
  apply: (
    sourceFile: ts.SourceFile,
    position: number,
    options: ts.GetCompletionsAtPositionOptions | undefined,
    formatCodeSettings: ts.FormatCodeSettings | undefined
  ) => Nano.Nano<
    Array<CompletionEntryDefinition>,
    never,
    | TypeCheckerApi.TypeCheckerApi
    | PluginOptions
    | TypeScriptApi.TypeScriptApi
    | TypeCheckerApi.TypeCheckerApiCache
  >
}

export interface CompletionEntryDefinition {
  name: string
  kind: ts.ScriptElementKind
  insertText: string
  isSnippet: true
  replacementSpan: ts.TextSpan
}

export function createCompletion(definition: CompletionDefinition): CompletionDefinition {
  return definition
}

export const PluginOptions = Nano.Tag<PluginOptions>("PluginOptions")

export class SourceFileNotFoundError extends Data.TaggedError("SourceFileNotFoundError")<{
  fileName: string
}> {}

export const getSemanticDiagnosticsWithCodeFixes = Nano.fn("LSP.getSemanticDiagnostics")(function*(
  rules: Array<DiagnosticDefinition>,
  sourceFile: ts.SourceFile
) {
  const effectDiagnostics: Array<ts.Diagnostic> = []
  const effectCodeFixes: Array<ApplicableDiagnosticDefinitionFixWithPositionAndCode> = []
  const executor = yield* createDiagnosticExecutor(sourceFile)
  for (const rule of rules) {
    const result = yield* (
      Nano.option(executor.execute(rule))
    )
    if (Option.isSome(result)) {
      effectDiagnostics.push(
        ...pipe(
          result.value,
          ReadonlyArray.map((_) => ({
            file: sourceFile,
            start: _.node.getStart(sourceFile),
            length: _.node.getEnd() - _.node.getStart(sourceFile),
            messageText: _.messageText,
            category: _.category,
            code: rule.code,
            source: "effect"
          }))
        )
      )
      effectCodeFixes.push(
        ...pipe(
          result.value,
          ReadonlyArray.map((_) =>
            ReadonlyArray.map(
              _.fixes,
              (fix) => ({
                ...fix,
                code: rule.code,
                start: _.node.getStart(sourceFile),
                end: _.node.getEnd()
              })
            )
          ),
          ReadonlyArray.flatten
        )
      )
    }
  }

  return ({
    diagnostics: effectDiagnostics,
    codeFixes: effectCodeFixes
  })
})

export const getApplicableRefactors = Nano.fn("LSP.getApplicableRefactors")(function*(
  refactors: Array<RefactorDefinition>,
  sourceFile: ts.SourceFile,
  positionOrRange: number | ts.TextRange
) {
  const textRange = typeof positionOrRange === "number"
    ? { pos: positionOrRange, end: positionOrRange }
    : positionOrRange
  const effectRefactors: Array<ts.ApplicableRefactorInfo> = []
  for (const refactor of refactors) {
    const result = yield* Nano.option(refactor.apply(sourceFile, textRange))
    if (Option.isSome(result)) {
      effectRefactors.push({
        name: refactor.name,
        description: refactor.description,
        actions: [{
          name: refactor.name,
          description: result.value.description,
          kind: result.value.kind
        }]
      })
    }
  }
  return effectRefactors
})

export const getEditsForRefactor = Nano.fn("LSP.getEditsForRefactor")(function*(
  refactors: Array<RefactorDefinition>,
  sourceFile: ts.SourceFile,
  positionOrRange: number | ts.TextRange,
  refactorName: string
) {
  const refactor = refactors.find((refactor) => refactor.name === refactorName)
  if (!refactor) {
    return yield* Nano.fail(new RefactorNotApplicableError())
  }
  const textRange = typeof positionOrRange === "number"
    ? { pos: positionOrRange, end: positionOrRange }
    : positionOrRange

  return yield* refactor.apply(sourceFile, textRange)
})

export const getCompletionsAtPosition = Nano.fn("LSP.getCompletionsAtPosition")(function*(
  completions: Array<CompletionDefinition>,
  sourceFile: ts.SourceFile,
  position: number,
  options: ts.GetCompletionsAtPositionOptions | undefined,
  formatCodeSettings: ts.FormatCodeSettings | undefined
) {
  const effectCompletions: Array<ts.CompletionEntry> = []
  for (const completion of completions) {
    const result = yield* completion.apply(sourceFile, position, options, formatCodeSettings)
    effectCompletions.push(
      ...result.map((_) => ({ sortText: "11", ..._ }) satisfies ts.CompletionEntry)
    )
  }
  return effectCompletions
})

const createDiagnosticExecutor = Nano.fn("LSP.createCommentDirectivesProcessor")(
  function*(sourceFile: ts.SourceFile) {
    const ts = yield* Nano.service(TypeScriptApi.TypeScriptApi)

    const ruleOverrides: Record<
      string,
      Array<{ start: number; end: number; level: string }>
    > = {}
    const skippedRules: Array<string> = []

    const regex =
      /@effect-diagnostics((?:\s[a-zA-Z0-9/]+:(?:off|warning|error|message|suggestion|skip-file))+)?/gm
    let match: RegExpExecArray | null
    while ((match = regex.exec(sourceFile.text)) !== null) {
      const rulesCaptureGroup = match[1]

      if (rulesCaptureGroup) {
        const trimmedRuleString = rulesCaptureGroup.trim()
        if (trimmedRuleString) {
          const individualRules = trimmedRuleString.split(/\s+/)
          for (const rulePair of individualRules) {
            const [ruleName, ruleLevel] = rulePair.toLowerCase().split(":")
            if (ruleName && ruleLevel) {
              if (ruleLevel === "skip-file") skippedRules.push(ruleName)
              ruleOverrides[ruleName] = ruleOverrides[ruleName] || []
              const newLength = ruleOverrides[ruleName].push({
                start: match.index,
                end: Number.MAX_SAFE_INTEGER,
                level: ruleLevel
              })
              if (newLength > 1) ruleOverrides[ruleName][newLength - 2].end = match.index
            }
          }
        }
      }
    }

    const levelToDiagnosticCategory: Record<string, ts.DiagnosticCategory> = {
      error: ts.DiagnosticCategory.Error,
      warning: ts.DiagnosticCategory.Warning,
      message: ts.DiagnosticCategory.Message,
      suggestion: ts.DiagnosticCategory.Suggestion
    }

    const execute = Nano.fn("LSP.ruleExecutor")(function*(
      rule: DiagnosticDefinition
    ) {
      const ruleNameLowered = rule.name.toLowerCase()
      // if file is skipped entirely, do not process the rule
      if (skippedRules.indexOf(ruleNameLowered) > -1) return []
      // run the executor
      let modifiedDiagnostics = yield* rule.apply(sourceFile)
      // apply overrides
      for (const override of (ruleOverrides[ruleNameLowered] || [])) {
        if (override.level === "off") {
          // remove from output those in range
          modifiedDiagnostics = modifiedDiagnostics.filter((_) =>
            !(_.node.getStart(sourceFile) >= override.start && _.node.getEnd() <= override.end)
          )
        } else {
          // change severity
          for (
            const message of modifiedDiagnostics.filter((_) =>
              _.node.getStart(sourceFile) >= override.start && _.node.getEnd() <= override.end
            )
          ) {
            message.category = override.level in levelToDiagnosticCategory
              ? levelToDiagnosticCategory[override.level]
              : message.category
          }
        }
      }

      // append a rule fix to disable this check for the entire file
      const fixByDisableEntireFile: ApplicableDiagnosticDefinitionFix = {
        fixName: rule.name + "_skipFile",
        description: "Disable " + rule.name + " for this file",
        apply: Nano.flatMap(
          Nano.service(TypeScriptApi.ChangeTracker),
          (changeTracker) =>
            Nano.sync(() =>
              changeTracker.insertText(
                sourceFile,
                0,
                `/** @effect-diagnostics ${rule.name}:skip-file */` + "\n"
              )
            )
        )
      }
      const rulesWithDisableFix = modifiedDiagnostics.map((diagnostic) => ({
        ...diagnostic,
        fixes: diagnostic.fixes.concat([fixByDisableEntireFile])
      }))

      return rulesWithDisableFix
    })

    return { execute }
  }
)
