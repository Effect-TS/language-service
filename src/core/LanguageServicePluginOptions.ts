import { isArray } from "effect/Array"
import * as Array from "effect/Array"
import { pipe } from "effect/Function"
import { hasProperty, isBoolean, isNumber, isObject, isRecord, isString } from "effect/Predicate"
import * as Record from "effect/Record"
import * as Nano from "./Nano"

export type DiagnosticSeverity = "error" | "warning" | "message" | "suggestion"

export interface LanguageServicePluginOptions {
  refactors: boolean
  diagnostics: boolean
  diagnosticSeverity: Record<string, DiagnosticSeverity | "off">
  quickinfoEffectParameters: "always" | "never" | "whentruncated"
  quickinfo: boolean
  quickinfoMaximumLength: number
  completions: boolean
  goto: boolean
  inlays: boolean
  allowedDuplicatedPackages: Array<string>
  namespaceImportPackages: Array<string>
  topLevelNamedReexports: "ignore" | "follow"
  barrelImportPackages: Array<string>
  importAliases: Record<string, string>
  renames: boolean
  noExternal: boolean
}

export const LanguageServicePluginOptions = Nano.Tag<LanguageServicePluginOptions>("PluginOptions")

function parseDiagnosticSeverity(config: Record<PropertyKey, unknown>): Record<string, DiagnosticSeverity | "off"> {
  if (!isRecord(config)) return {}
  return Object.fromEntries(
    pipe(
      Object.entries(config),
      Array.filter(([key, value]) => isString(key) && isString(value)),
      Array.map(([key, value]) => [String(key).toLowerCase(), String(value).toLowerCase()]),
      Array.filter(([_, value]) =>
        value === "off" || value === "error" || value === "warning" || value === "message" || value === "suggestion"
      )
    )
  )
}

export const defaults: LanguageServicePluginOptions = {
  refactors: true,
  diagnostics: true,
  diagnosticSeverity: {},
  quickinfo: true,
  quickinfoEffectParameters: "whentruncated",
  quickinfoMaximumLength: -1,
  completions: true,
  goto: true,
  inlays: true,
  allowedDuplicatedPackages: [],
  namespaceImportPackages: [],
  topLevelNamedReexports: "ignore",
  barrelImportPackages: [],
  importAliases: {},
  renames: true,
  noExternal: false
}

export function parse(config: any): LanguageServicePluginOptions {
  return {
    refactors: isObject(config) && hasProperty(config, "refactors") && isBoolean(config.refactors)
      ? config.refactors
      : defaults.refactors,
    diagnostics: isObject(config) && hasProperty(config, "diagnostics") && isBoolean(config.diagnostics)
      ? config.diagnostics
      : defaults.diagnostics,
    diagnosticSeverity:
      isObject(config) && hasProperty(config, "diagnosticSeverity") && isRecord(config.diagnosticSeverity)
        ? parseDiagnosticSeverity(config.diagnosticSeverity)
        : defaults.diagnosticSeverity,
    quickinfo: isObject(config) && hasProperty(config, "quickinfo") && isBoolean(config.quickinfo)
      ? config.quickinfo
      : defaults.quickinfo,
    quickinfoEffectParameters: isObject(config) && hasProperty(config, "quickinfoEffectParameters") &&
        isString(config.quickinfoEffectParameters) &&
        ["always", "never", "whentruncated"].includes(config.quickinfoEffectParameters.toLowerCase())
      ? config.quickinfoEffectParameters.toLowerCase() as "always" | "never" | "whentruncated"
      : defaults.quickinfoEffectParameters,
    quickinfoMaximumLength:
      isObject(config) && hasProperty(config, "quickinfoMaximumLength") && isNumber(config.quickinfoMaximumLength)
        ? config.quickinfoMaximumLength
        : defaults.quickinfoMaximumLength,
    completions: isObject(config) && hasProperty(config, "completions") && isBoolean(config.completions)
      ? config.completions
      : defaults.completions,
    goto: isObject(config) && hasProperty(config, "goto") && isBoolean(config.goto)
      ? config.goto
      : defaults.goto,
    inlays: isObject(config) && hasProperty(config, "inlays") && isBoolean(config.inlays)
      ? config.inlays
      : defaults.inlays,
    allowedDuplicatedPackages: isObject(config) && hasProperty(config, "allowedDuplicatedPackages") &&
        isArray(config.allowedDuplicatedPackages) && config.allowedDuplicatedPackages.every(isString)
      ? config.allowedDuplicatedPackages.map((_) => _.toLowerCase())
      : defaults.allowedDuplicatedPackages,
    namespaceImportPackages: isObject(config) && hasProperty(config, "namespaceImportPackages") &&
        isArray(config.namespaceImportPackages) && config.namespaceImportPackages.every(isString)
      ? config.namespaceImportPackages.map((_) => _.toLowerCase())
      : defaults.namespaceImportPackages,
    barrelImportPackages: isObject(config) && hasProperty(config, "barrelImportPackages") &&
        isArray(config.barrelImportPackages) && config.barrelImportPackages.every(isString)
      ? config.barrelImportPackages.map((_) => _.toLowerCase())
      : defaults.barrelImportPackages,
    importAliases: isObject(config) && hasProperty(config, "importAliases") && isRecord(config.importAliases)
      ? Record.map(config.importAliases, (value) => String(value))
      : defaults.importAliases,
    topLevelNamedReexports: isObject(config) && hasProperty(config, "topLevelNamedReexports") &&
        isString(config.topLevelNamedReexports) &&
        ["ignore", "follow"].includes(config.topLevelNamedReexports.toLowerCase())
      ? config.topLevelNamedReexports.toLowerCase() as "ignore" | "follow"
      : defaults.topLevelNamedReexports,
    renames: isObject(config) && hasProperty(config, "renames") && isBoolean(config.renames)
      ? config.renames
      : defaults.renames,
    noExternal: isObject(config) && hasProperty(config, "noExternal") && isBoolean(config.noExternal)
      ? config.noExternal
      : defaults.noExternal
  }
}
