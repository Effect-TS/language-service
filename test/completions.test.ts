import { completions } from "@effect/language-service/completions"
import * as LanguageServicePluginOptions from "@effect/language-service/core/LanguageServicePluginOptions"
import * as LSP from "@effect/language-service/core/LSP"
import * as Nano from "@effect/language-service/core/Nano"
import * as TypeCheckerApi from "@effect/language-service/core/TypeCheckerApi"
import * as TypeParser from "@effect/language-service/core/TypeParser"
import * as TypeScriptApi from "@effect/language-service/core/TypeScriptApi"
import * as TypeScriptUtils from "@effect/language-service/core/TypeScriptUtils"
import * as Either from "effect/Either"
import { pipe } from "effect/Function"
import * as fs from "fs"
import * as path from "path"
import * as ts from "typescript"
import { describe, expect, it } from "vitest"
import { createServicesWithMockedVFS } from "./utils/mocks.js"

const getExamplesCompletionsDir = () => path.join(__dirname, "..", "examples", "completions")

function testCompletionOnExample(
  completion: LSP.CompletionDefinition,
  fileName: string,
  sourceText: string,
  textRangeString: string
) {
  const { program, sourceFile } = createServicesWithMockedVFS(
    fileName,
    sourceText
  )

  // gets the position to test
  let startPos = 0
  for (const lineAndCol of textRangeString.split("-")) {
    const [line, character] = lineAndCol.split(":")
    startPos = ts.getPositionOfLineAndCharacter(sourceFile, +line! - 1, +character! - 1)
  }

  // check and assert the completions is executable
  const maybeEntries = pipe(
    LSP.getCompletionsAtPosition(
      [completion],
      sourceFile,
      startPos,
      undefined,
      ts.getDefaultFormatCodeSettings("\n")
    ),
    TypeParser.nanoLayer,
    TypeScriptUtils.nanoLayer,
    Nano.provideService(TypeCheckerApi.TypeCheckerApi, program.getTypeChecker()),
    Nano.provideService(TypeScriptApi.TypeScriptProgram, program),
    Nano.provideService(TypeScriptApi.TypeScriptApi, ts),
    Nano.provideService(LanguageServicePluginOptions.LanguageServicePluginOptions, {
      diagnostics: false,
      diagnosticSeverity: {},
      quickinfo: false,
      completions: false,
      goto: false,
      allowedDuplicatedPackages: [],
      namespaceImportPackages: [],
      barrelImportPackages: [],
      topLevelNamedReexports: "ignore"
    }),
    Nano.unsafeRun
  )

  if (Either.isLeft(maybeEntries)) {
    expect(sourceText).toMatchSnapshot()
    return
  }

  expect(maybeEntries.right).toMatchSnapshot()
}

function testAllCompletions() {
  // read all filenames
  const allExampleFiles = fs.readdirSync(getExamplesCompletionsDir())
  // for each definition
  for (const completion of completions) {
    // all files that start with the name and end with .ts
    const exampleFiles = allExampleFiles.filter((fileName) =>
      fileName === completion.name + ".ts" ||
      fileName.startsWith(completion.name + "_") && fileName.endsWith(".ts")
    )
    describe("Completion " + completion.name, () => {
      // for each example file
      for (const fileName of exampleFiles) {
        // first we extract from the first comment line all the positions where the refactor has to be tested
        const sourceWithMarker = fs.readFileSync(path.join(getExamplesCompletionsDir(), fileName))
          .toString("utf8")
        const firstLine = (sourceWithMarker.split("\n")[0] || "").trim()
        for (const [textRangeString] of firstLine.matchAll(/([0-9]+:[0-9]+(-[0-9]+:[0-9]+)*)/gm)) {
          it(fileName + " at " + textRangeString, () => {
            // create the language service
            const sourceText = "// Result of running completion " + completion.name +
              " at position " + textRangeString + sourceWithMarker.substring(firstLine.length)
            testCompletionOnExample(completion, fileName, sourceText, textRangeString)
          })
        }
      }
    })
  }
}

testAllCompletions()
