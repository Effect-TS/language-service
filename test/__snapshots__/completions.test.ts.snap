// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Completion contextSelfInClasses > contextSelfInClasses.ts at 4:40 1`] = `
[
  {
    "insertText": "Context.Tag("MyService")<MyService, \${0}>(){}",
    "isSnippet": true,
    "kind": "const",
    "name": "Tag("MyService")",
    "replacementSpan": {
      "length": 8,
      "start": 144,
    },
    "sortText": "11",
  },
]
`;

exports[`Completion durationInput > durationInput.ts at 4:38 1`] = `
[
  {
    "insertText": "\${0} nanos",
    "isSnippet": true,
    "kind": "string",
    "name": "nanos",
    "sortText": "11",
  },
  {
    "insertText": "\${0} micros",
    "isSnippet": true,
    "kind": "string",
    "name": "micros",
    "sortText": "11",
  },
  {
    "insertText": "\${0} millis",
    "isSnippet": true,
    "kind": "string",
    "name": "millis",
    "sortText": "11",
  },
  {
    "insertText": "\${0} seconds",
    "isSnippet": true,
    "kind": "string",
    "name": "seconds",
    "sortText": "11",
  },
  {
    "insertText": "\${0} minutes",
    "isSnippet": true,
    "kind": "string",
    "name": "minutes",
    "sortText": "11",
  },
  {
    "insertText": "\${0} hours",
    "isSnippet": true,
    "kind": "string",
    "name": "hours",
    "sortText": "11",
  },
  {
    "insertText": "\${0} days",
    "isSnippet": true,
    "kind": "string",
    "name": "days",
    "sortText": "11",
  },
  {
    "insertText": "\${0} weeks",
    "isSnippet": true,
    "kind": "string",
    "name": "weeks",
    "sortText": "11",
  },
]
`;

exports[`Completion durationInput > durationInput_props.ts at 5:14 1`] = `
[
  {
    "insertText": "\${0} nanos",
    "isSnippet": true,
    "kind": "string",
    "name": "nanos",
    "sortText": "11",
  },
  {
    "insertText": "\${0} micros",
    "isSnippet": true,
    "kind": "string",
    "name": "micros",
    "sortText": "11",
  },
  {
    "insertText": "\${0} millis",
    "isSnippet": true,
    "kind": "string",
    "name": "millis",
    "sortText": "11",
  },
  {
    "insertText": "\${0} seconds",
    "isSnippet": true,
    "kind": "string",
    "name": "seconds",
    "sortText": "11",
  },
  {
    "insertText": "\${0} minutes",
    "isSnippet": true,
    "kind": "string",
    "name": "minutes",
    "sortText": "11",
  },
  {
    "insertText": "\${0} hours",
    "isSnippet": true,
    "kind": "string",
    "name": "hours",
    "sortText": "11",
  },
  {
    "insertText": "\${0} days",
    "isSnippet": true,
    "kind": "string",
    "name": "days",
    "sortText": "11",
  },
  {
    "insertText": "\${0} weeks",
    "isSnippet": true,
    "kind": "string",
    "name": "weeks",
    "sortText": "11",
  },
]
`;

exports[`Completion effectDataClasses > effectDataClasses.ts at 4:35 1`] = `
[
  {
    "insertText": "Data.TaggedError("MyError")<{\${0}}>{}",
    "isSnippet": true,
    "kind": "const",
    "name": "TaggedError("MyError")",
    "replacementSpan": {
      "length": 5,
      "start": 133,
    },
    "sortText": "11",
  },
  {
    "insertText": "Data.TaggedClass("MyError")<{\${0}}>{}",
    "isSnippet": true,
    "kind": "const",
    "name": "TaggedClass("MyError")",
    "replacementSpan": {
      "length": 5,
      "start": 133,
    },
    "sortText": "11",
  },
]
`;

exports[`Completion effectDiagnosticsComment > effectDiagnosticsComment.ts at 2:5 1`] = `
[
  {
    "insertText": "@effect-diagnostics \${1|duplicatePackage,effectInVoidSuccess,floatingEffect,genericEffectServices,importFromBarrel,leakingRequirements,missingEffectContext,missingEffectError,missingReturnYieldStar,missingStarInYieldEffectGen,returnEffectInGen,scopeInLayerEffect,unnecessaryEffectGen,unnecessaryPipe|}:\${2|off,warning,error,message,suggestion|}$0",
    "isSnippet": true,
    "kind": "string",
    "name": "@effect-diagnostics",
    "replacementSpan": {
      "length": 1,
      "start": 76,
    },
    "sortText": "11",
  },
  {
    "insertText": "@effect-diagnostics-next-line \${1|duplicatePackage,effectInVoidSuccess,floatingEffect,genericEffectServices,importFromBarrel,leakingRequirements,missingEffectContext,missingEffectError,missingReturnYieldStar,missingStarInYieldEffectGen,returnEffectInGen,scopeInLayerEffect,unnecessaryEffectGen,unnecessaryPipe|}:\${2|off,warning,error,message,suggestion|}$0",
    "isSnippet": true,
    "kind": "string",
    "name": "@effect-diagnostics-next-line",
    "replacementSpan": {
      "length": 1,
      "start": 76,
    },
    "sortText": "11",
  },
]
`;

exports[`Completion effectSchemaSelfInClasses > effectSchemaSelfInClasses_dotToken.ts at 4:22 1`] = `
[
  {
    "insertText": "S.Class<Test>("Test")({\${0}}){}",
    "isSnippet": true,
    "kind": "const",
    "name": "Class<Test>",
    "replacementSpan": {
      "length": 2,
      "start": 130,
    },
    "sortText": "11",
  },
  {
    "insertText": "S.TaggedError<Test>("Test")("Test", {\${0}}){}",
    "isSnippet": true,
    "kind": "const",
    "name": "TaggedError<Test>",
    "replacementSpan": {
      "length": 2,
      "start": 130,
    },
    "sortText": "11",
  },
  {
    "insertText": "S.TaggedClass<Test>("Test")("Test", {\${0}}){}",
    "isSnippet": true,
    "kind": "const",
    "name": "TaggedClass<Test>",
    "replacementSpan": {
      "length": 2,
      "start": 130,
    },
    "sortText": "11",
  },
  {
    "insertText": "S.TaggedRequest<Test>("Test")("Test", {\${0}}){}",
    "isSnippet": true,
    "kind": "const",
    "name": "TaggedRequest<Test>",
    "replacementSpan": {
      "length": 2,
      "start": 130,
    },
    "sortText": "11",
  },
]
`;

exports[`Completion effectSchemaSelfInClasses > effectSchemaSelfInClasses_tagg.ts at 4:25 1`] = `
[
  {
    "insertText": "S.Class<Test>("Test")({\${0}}){}",
    "isSnippet": true,
    "kind": "const",
    "name": "Class<Test>",
    "replacementSpan": {
      "length": 5,
      "start": 130,
    },
    "sortText": "11",
  },
  {
    "insertText": "S.TaggedError<Test>("Test")("Test", {\${0}}){}",
    "isSnippet": true,
    "kind": "const",
    "name": "TaggedError<Test>",
    "replacementSpan": {
      "length": 5,
      "start": 130,
    },
    "sortText": "11",
  },
  {
    "insertText": "S.TaggedClass<Test>("Test")("Test", {\${0}}){}",
    "isSnippet": true,
    "kind": "const",
    "name": "TaggedClass<Test>",
    "replacementSpan": {
      "length": 5,
      "start": 130,
    },
    "sortText": "11",
  },
  {
    "insertText": "S.TaggedRequest<Test>("Test")("Test", {\${0}}){}",
    "isSnippet": true,
    "kind": "const",
    "name": "TaggedRequest<Test>",
    "replacementSpan": {
      "length": 5,
      "start": 130,
    },
    "sortText": "11",
  },
]
`;

exports[`Completion effectSelfInClasses > effectSelfInClasses.ts at 4:39 1`] = `
[
  {
    "insertText": "Effect.Service<MyService>()("MyService", {\${0}}){}",
    "isSnippet": true,
    "kind": "const",
    "name": "Service<MyService>",
    "replacementSpan": {
      "length": 7,
      "start": 141,
    },
    "sortText": "11",
  },
]
`;

exports[`Completion fnFunctionStar > fnFunctionStar.ts at 4:24 1`] = `
[
  {
    "insertText": "fn("program")(function*(\${1}){\${0}})",
    "isSnippet": true,
    "kind": "const",
    "name": "fn("program")",
    "replacementSpan": {
      "length": 0,
      "start": 128,
    },
    "sortText": "11",
  },
  {
    "insertText": "fn(function*(\${1}){\${0}})",
    "isSnippet": true,
    "kind": "const",
    "name": "fn(function*(){})",
    "replacementSpan": {
      "length": 0,
      "start": 128,
    },
    "sortText": "11",
  },
  {
    "insertText": "fnUntraced(function*(\${1}){\${0}})",
    "isSnippet": true,
    "kind": "const",
    "name": "fnUntraced(function*(){})",
    "replacementSpan": {
      "length": 0,
      "start": 128,
    },
    "sortText": "11",
  },
]
`;

exports[`Completion genFunctionStar > genFunctionStar.ts at 4:24 1`] = `
[
  {
    "insertText": "gen(function*(){\${0}})",
    "isSnippet": true,
    "kind": "const",
    "name": "gen(function*(){})",
    "replacementSpan": {
      "length": 0,
      "start": 129,
    },
    "sortText": "11",
  },
]
`;

exports[`Completion rpcMakeClasses > rpcMakeClasses.ts at 4:36 1`] = `
[
  {
    "insertText": "Rpc.make("MyRequest", {\${0}}) {}",
    "isSnippet": true,
    "kind": "const",
    "name": "make("MyRequest")",
    "replacementSpan": {
      "length": 4,
      "start": 135,
    },
    "sortText": "11",
  },
]
`;
